File: frontend/src/domain/repositories/product.repository.ts
/**
 * Product Repository Interface
 * 완제품 데이터 접근 인터페이스
 */

import type {
  Product,
  ProductCreateInput,
  ProductUpdateInput,
  ProductStatus,
} from '../entities/product';
import type { BaseRepository } from './base.repository';

export interface IProductRepository
  extends BaseRepository<Product, ProductCreateInput, ProductUpdateInput> {
  /**
   * 상태별 조회
   */
  getByStatus(status: ProductStatus): Promise<Product[]>;

  /**
   * 고객사별 조회
   */
  getByCustomer(customerName: string): Promise<Product[]>;

  /**
   * 완제품 검색 (품번, 품명, 고객품번)
   */
  search(query: string): Promise<Product[]>;
}

File: backend/backend/domain/repositories/product_repository.py
"""ProductRepository 인터페이스."""

from __future__ import annotations

from abc import abstractmethod

from backend.domain.entities.product import Product, ProductStatus
from backend.domain.repositories.base import BaseRepository


class ProductRepository(BaseRepository[Product, str]):
    """완제품 Repository 인터페이스."""

    @abstractmethod
    async def get_by_status(self, status: ProductStatus) -> list[Product]:
        """상태별 완제품 조회."""
        ...

    @abstractmethod
    async def get_by_customer(self, customer: str) -> list[Product]:
        """고객사별 완제품 조회."""
        ...
